src/throttler/throttler.ts

```diff
@@ -254,4 +230,79 @@ export abstract class Throttler<T, R> {
       return p.resolve();
     });
   }
+
+  private initializeTimeout(taskData: TaskData<T, R>, cursorIndex: number): Promise<void> {
+    let rejectFunc: (err: TaskError) => void;
+    const timeoutMillis = taskData.timeoutMillis!;
+    const timeoutPromise = new Promise<void>((_, reject) => {
+      rejectFunc = reject;
+    });
+    taskData.timeoutId = setTimeout(() => {
+      taskData.isTimedOut = true;
+      rejectFunc(new TimeoutError(this.taskName(cursorIndex), timeoutMillis));
+    }, timeoutMillis);
+
+    return timeoutPromise;
+  }
+
+  private async executeTask(taskData: TaskData<T, R>, cursorIndex: number): Promise<any> {
+    const t0 = Date.now();
+    try {
+      const result = await this.handler(taskData.task);
+      const dt = Date.now() - t0;
+      if (dt < this.min) {
+        this.min = dt;
+      }
+      if (dt > this.max) {
+        this.max = dt;
+      }
+      this.avg = (this.avg * this.complete + dt) / (this.complete + 1);
+
+      this.success++;
+      this.complete++;
+      this.active--;
+      return result;
+    } catch (error) {
+      if (taskData.retryCount === this.retries) {
+        throw new RetriesExhaustedError(this.taskName(cursorIndex), this.retries, error);
+      }
+      if (taskData.isTimedOut) {
+        throw error;
+      }
+      this.retried++;
+      taskData.retryCount++;
+      await backoff(taskData.retryCount, this.backoff);
```

Comment:
It's actually more important to check `taskData.isTimedOut` *after* `backoff`, since running `this.executeTask` is potentially much more expensive than `backoff`. I suppose, ideally, we should check for the timeout both before and after `backoff`.

Reply:
This is actually one of the problem that I thought of. I agree that it's best to check both before and after `backoff`

Reply:
I'm thinking about not using `isTimedOut`, instead we store `timeoutPromise` in `TaskData`, we can `race` the `backoffPromise` and `timeoutPromise`. This way the duplicate code can be avoided. But other than more clean code, I don't see any other advantage:
```typescript
...
await Promise.race([taskData.timeoutPromise, await backoff(taskData.retryCount + 1, this.backoff)]);
this.retried++;
taskData.
logger.debug(`[${this.name}] Retrying task`, this.taskName(cursorIndex));
return this.executeTask(taskData, cursorIndex);

Reply:
Pushed an update that only throws `Timeout` after backoff

Reply:
Looks good